packages/react-pages/src/client/useAppState.tsx

```diff
@@ -1,90 +1,54 @@
-    if (
-      newLoadState.type === loadState.type &&
-      newLoadState.routePath === loadState.routePath
-    ) {
-      // don't set state if prev is already what I want
-    } else {
-      _setLoadState(newLoadState)
-    }
-  }
+  const [loadState, setLoadState] = useState<LoadState>(() => ({
+    type: 'loading',
+    routePath,
+  }))
 
-  const { routePath } = loadState
+  const onLoadState = (
+    type: LoadState['type'],
+    routePath: string,
+    error?: any
+  ) =>
+    (type !== loadState.type || routePath !== loadState.routePath) &&
+    setLoadState({ type, routePath, error })
 
-  if (!pages[latestRoutePath]) {
-    if (pages['/404'] && !dataCache['/404']) {
-      // load /404 page
-      setLoadState({
-        type: 'loading',
-        routePath: latestRoutePath,
-      })
+  const loading = usePageModule(routePath)
+  const loadingRef = useRef<Promise<any> | undefined>()
+  useLayoutEffect(() => {
+    loadingRef.current = loading
+    if (!loading) {
+      onLoadState('404', routePath)
     } else {
-      setLoadState({
-        type: '404',
-        routePath: latestRoutePath,
-      })
-    }
-  } else if (routePath !== latestRoutePath) {
-    // if routePath has changed,
-    // update loadState and rerender this component
-    setLoadState({
-      type: 'loading',
-      routePath: latestRoutePath,
-    })
-  } else if (dataCache[routePath]) {
-    // if we have the data in cache (.e.g during ssr or loaded done)
-    // update loadState and rerender this component
-    setLoadState({
-      type: 'loaded',
-      routePath,
-    })
-  }
-
-  useEffect(() => {
-    // FIXME handle race condition of this async setState
-    if (loadState.type === 'loading') {
-      let loadPath = routePath
-      if (!pages[routePath]) {
-        // loading a non-exist page
-        if (pages['/404']) {
-          loadPath = '/404'
-        }
+      if (dataCache[routePath]) {
+        onLoadState('loaded', routePath)
+      } else {
+        onLoadState('loading', routePath)
+        loading.then(
+          (page) =>
+            loading === loadingRef.current &&
+            batchedUpdates(() => {
+              onLoadState('loaded', routePath)
+              setDataCache((prev) => ({
+                ...prev,
+                [routePath]: page.default,
+              }))
+            }),
+          (error) =>
+            loading === loadingRef.current &&
+            onLoadState('load-error', routePath, error)
+        )
       }
-      const { data: dataImporter } = pages[loadPath]
-      dataImporter()
-        .then(({ default: pageLoaded }) => {
-          setDataCache((prev) => ({
-            ...prev,
-            [loadPath]: pageLoaded,
-          }))
-        })
-        .catch((error) => {
-          setLoadState({
-            type: 'load-error',
-            routePath,
-            error,
-          })
-          throw error
-        })
     }
-  }, [routePath])
+  }, [loading])
```

Comment:
Should we put `onLoadState` in the deps arrary?

Reply:
I don't think so. We only want the effect to run when `loading` changes.